Single linked list

Robin-Bjornsdotter opened the pull request.

Commit:
Attempt to make create, free, append and pop work.

Commit:
Added more append/pop options

Commit:
Little fixup in dynamic array allocation (array_2d)

Commit:
Fixed list_append_element and list_append_many testcase

Commit:
Fixed list_pop & added testcase call

disconnect3d commented on linked_list/single_linked_list.c:
```diff
@@ -0,0 +1,165 @@
+  return result;
+}
+
+void list_append_element_beginning(node* list, int new_element){
```
This is not used anywhere, prepare a testcase for it and ... try to understand why it doesn't work as you expected.

disconnect3d commented on linked_list/single_linked_list.c:
```diff
@@ -0,0 +1,165 @@
+  new_node->data = new_element;
+  new_node->next = NULL;
+
+  if (list->next == NULL)
```
Here's an undefined behavior (UB) when `list == NULL`.

disconnect3d replied:
Also, can be simplified to:

```
void list_append_element_end(node* list, int new_element){
  node* new_node = (node*)malloc(sizeof(node)); 
  CHECK_ALLOC(new_node); 
  new_node->data = new_element;
  new_node->next = NULL;

  while (list->next != NULL)
      list = list->next;

  list->next = new_node;
}
```

(This solution still doesn't fix the problem with `list==NULL`)

disconnect3d commented on linked_list/single_linked_list.c:
```diff
@@ -0,0 +1,165 @@
+
+
+void list_free(node* first){
+  struct node* tmp;
```
This can be moved to the scope of `while` loop.
(It is a good practice to narrow the scope of variable, so it will be harder to make a mistake and easier to follow the code)

disconnect3d commented on linked_list/single_linked_list.c:
```diff
@@ -0,0 +1,165 @@
+  }
+}
+
+node* list_pop_beginning(node* list){
```
This is not used anywhere, prepare a testcase for it and ... try to understand why it doesn't work as you expected.

disconnect3d commented on linked_list/single_linked_list.c:
```diff
@@ -0,0 +1,165 @@
+  return list;
+}
+
+void list_append_element_x(node* list, int new_element, int position){
```
Not used anywhere / prepare a testcase.
Also if `position == 0` the `left` is not initialized, therefore the `left->next = new_node` assignment is UB.

disconnect3d commented on linked_list/single_linked_list.c:
```diff
@@ -0,0 +1,165 @@
+  new_node->next = right;
+}
+
+void list_pop_element_x(node* list, int position){
```
Not used anywhere / prepare a testcase.
Also `if position == 0` the left is not initialized, therefore the `left->next = right->next` assignment is UB.

disconnect3d commented on linked_list/single_linked_list.c:
```diff
@@ -0,0 +1,165 @@
+}
+
+int main(void) {
+  
```
Fix indents here (either use 2 in all the code or 4).

Commit:
Added missing testcases

deleted append/pop on begginning and an attempt to fix append/pop on
chosen position

disconnect3d commented on linked_list/single_linked_list.c:
```diff
@@ -0,0 +1,186 @@
+    list = list->next;
+
+  list->next = new_node;
+}
```
1) `list_append_element_end` is pretty long as for a function (sometimes very long function names are okay, e.g. for test functions) - I would name it just `list_append`, as `append` is known to add to an end (e.g. `.append()` method in Python).
2) This won't work when `list == NULL`. Add such a test and try to find out what's wrong:

```
void test_append_to_empty() {
    node* head = NULL;
    list_append_element_end(head, 1);
    ASSERT(head != NULL);
    ASSERT(head->data == 1);
    ASSERT(head->next == NULL);
    list_free(head);
}
```

disconnect3d commented on linked_list/single_linked_list.c:
```diff
@@ -0,0 +1,186 @@
+  int result = list->data; 
+  free(list);
+  return result;
+}
```
I'd name it `list_pop`.

Some references:
- https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Array/pop
- https://docs.python.org/3/tutorial/datastructures.html#more-on-lists

Also... <strike>you don't have to fix that, but</strike> keep in mind this booms when `list == NULL`... (it doesn't handle such edge case).
It could either set a global variable that would say "it errored" (such global would have to be cleared by the user before calling it) (but globals are usually bad... also it wouldn't work with threads) or just return `-1` when `list == NULL` (and state it in some docstring/comment).

disconnect3d commented on linked_list/single_linked_list.c:
```diff
@@ -0,0 +1,186 @@
+      left->next = new_node;
+      new_node->next = right;
+    }
+  }
```
1) I would rename to `list_insert`.
2) Same bug as in `list_append_element_end` - won't work with `list == NULL`. Create a specific testcase for it.

disconnect3d commented on linked_list/single_linked_list.c:
```diff
@@ -0,0 +1,186 @@
+      free(right);
+    }
+    return popped_el;
+  }
```
1) Would rename to `list_remove`.
2) Booms with `list == NULL`.
3) Works weird with `position == 0`, as after such code:

```
node* head = create_list(0);
list_pop_on_x(head, 0);
```

head points to a memory which has been deallocated - it is easy to make a mistake right now.

disconnect3d commented on linked_list/single_linked_list.c:
```diff
@@ -0,0 +1,186 @@
+      list_append_on_x(list, 16, 2);
+      ASSERT(list->next->data == 16);
+      list_free(list);
+    }
```
If `position` is the same thing as `index` (so `0` means first node, `1` means second and so on), the assertion should be `list->next->next->data == 16`.

(I hadn't checked the implementations fully as there are some major bugs like unhandled testcases // pointed it out in the other comments)

Commit:
Fixed testcases with NULL list, renamed functions

disconnect3d commented on linked_list/single_linked_list.c:
```diff
@@ -0,0 +1,228 @@
+  test_list_create();
+  test_list_append();
+  test_list_append_many(); 
+  test_append_to_empty();
```
All the tests starts with `test_list_` except one ;).

disconnect3d commented on linked_list/single_linked_list.c:
```diff
@@ -0,0 +1,228 @@
+	new_node->next = right;
+      }
+  }
+}
```
The `list_insert` will boom with such testcase:

``` C
node* head = create_list(1);
list_insert(&head, 2, 10);
```

disconnect3d commented on linked_list/single_linked_list.c:
```diff
@@ -0,0 +1,228 @@
+      for (int i = 0; i < position; ++i){
+	left = right;
+	right = right->next;
+      }
```
The `list_remove` will boom in such case:

``` C
node* head = create_list(1);
list_remove(&head, 10);
```

Commit:
Changed list_insert and list_free

Commit:
Added list_reverse & makefile

disconnect3d commented on linked_list/single_linked_list.c:
```diff
@@ -0,0 +1,311 @@
+#include <stdio.h>
+#include <stdlib.h>
+#define ASSERT(x) printf("Assertion: " #x " = %d\n", (x))
```
As the assert is a macro, make it print the current function name (`__func__`) and line number (`__LINE__`) (without it, it could have been a function).

Also, make it print only when the condition is not met, so maybe:

``` C
if (!x) printf(...)
```

disconnect3d commented on linked_list/single_linked_list.c:
```diff
@@ -0,0 +1,311 @@
+    result = (*list)->data; 
+    free(*list);
+    return result;
+  }
```
Lets change those `*list` to a `node* tmp = *list;` variable. It will be more readable :F.

disconnect3d replied:
Also add a comment above the function, about "how it fails".

(Ofc in a real implementation it would have to return some kind of error, because currently it works similar to `atoi`).

disconnect3d commented on linked_list/single_linked_list.c:
```diff
@@ -0,0 +1,311 @@
+	  node* remove = temp->next;
+	  temp->next = NULL;
+	  free(remove);
+	}
```
Booms on:

``` C
node* head = list_create(0);
list_remove(&head, 5);
```

Basically this:

``` C
while (temp->next) { temp = temp->next; ... }
if (temp->next->next == NULL) { ... }
```

Looks scary no matter of the conditional `break` instruction happening.
(If `break` occurs, the `temp->next` might be `NULL`. If `break` doesn't occur, the `temp->next` is `NULL`).

Commit:
Attempt to fix list_remove

Commit:
New assertion

Commit:
Fixed list_remove

Commit:
Changed indentation in list_remove
